File: ngff_zarr/ngff_image.py
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Mapping, Optional, Sequence

from dask.array.core import Array as DaskArray

from .zarr_metadata import Units

ComputedCallback = Callable[[], None]


@dataclass
class NgffImage:
    data: DaskArray
    dims: Sequence[str]
    scale: Dict[str, float]
    translation: Dict[str, float]
    name: str = "image"
    axes_units: Optional[Mapping[str, Units]] = None
    computed_callbacks: List[ComputedCallback] = field(default_factory=list)

File: ts/src/schemas/ngff_image.ts
import { z } from "zod";
import { UnitsSchema } from "./units.ts";

// Since NgffImage now uses zarr.Array directly, we create a basic schema for validation
const ZarrArraySchema = z.object({
  shape: z.array(z.number()),
  dtype: z.string(),
  chunks: z.array(z.number()),
  path: z.string().optional(),
  name: z.string().optional(),
});

export const NgffImageOptionsSchema: z.ZodObject<{
  data: typeof ZarrArraySchema;
  dims: z.ZodArray<z.ZodString>;
  scale: z.ZodRecord<z.ZodString, z.ZodNumber>;
  translation: z.ZodRecord<z.ZodString, z.ZodNumber>;
  name: z.ZodDefault<z.ZodString>;
  axesUnits: z.ZodOptional<z.ZodRecord<z.ZodString, typeof UnitsSchema>>;
  computedCallbacks: z.ZodDefault<z.ZodArray<z.ZodAny>>;
}> = z.object({
  data: ZarrArraySchema,
  dims: z.array(z.string()),
  scale: z.record(z.string(), z.number()),
  translation: z.record(z.string(), z.number()),
  name: z.string().default("image"),
  axesUnits: z.record(z.string(), UnitsSchema).optional(),
  computedCallbacks: z.array(z.any()).default([]),
});
